### PR TITLE
Integrate AIDE ML code optimization

### PR DESCRIPTION
**Pull request by efsiatras**
## Description

This PR adds a new `oumi aide` command that brings agentic code optimization to Oumi using [AIDE](https://github.com/WecoAI/aideml), the open-source ML engineering agent by [Weco AI](https://www.weco.ai/).

### Why AIDE

Unlike `oumi tune` which searches a predefined hyperparameter space, AIDE uses an LLM to write, execute, and iteratively improve Python code through tree search (draft → debug → improve). It can explore code-level changes like reward function design, evaluation logic, training strategies.

AIDE implements the algorithm from [arXiv:2502.13138](https://arxiv.org/abs/2502.13138) and has been independently validated by OpenAI ([MLE-bench](https://arxiv.org/abs/2410.07095), 4x more medals than the best linear agent across 75 Kaggle competitions), Meta ([LLM Speedrunning](https://arxiv.org/abs/2506.22419), [AI Research Agents](https://arxiv.org/abs/2507.02554)), Sakana AI ([AI Scientist-v2](https://arxiv.org/abs/2504.08066)), and METR ([RE-Bench](https://arxiv.org/abs/2411.15114)).

This PR integrates the open-source research version (`aideml`) which runs fully local. Weco AI also offers a separate [production platform](https://www.weco.ai/) with experiment tracking and cloud-hybrid architecture.

### Architecture

The integration mirrors the existing Optuna pattern:

- `core/configs/params/aide_params.py` → AideParams (mirrors TuningParams)
- `core/configs/aide_config.py` → AideConfig (mirrors TuningConfig)
- `core/agentic/base_agentic_optimizer.py` → BaseAgenticOptimizer ABC (mirrors BaseTuner)
- `core/agentic/aide_optimizer.py` → AideOptimizer (mirrors OptunaTuner)
- `core/agentic/workspace_helper.py` → Helper script injected into workspace
- `builders/agentic.py` → build_agentic_optimizer() (mirrors build_tuner)
- `aide.py` → Orchestration (mirrors tune.py)
- `cli/aide_cmd.py` → CLI command (mirrors cli/tune.py)

### Optimization surfaces

- **CONFIG_SEARCH** — modifies training hyperparameters (learning rate, optimizer, LoRA rank, etc.)
- **REWARD_FUNCTION** — designs reward functions for GRPO/RLHF
- **EVAL_FUNCTION** — generates custom evaluation functions
- **FULL_PIPELINE** — writes complete training scripts with no constraints

### Workspace helper

Instead of exposing raw Oumi APIs to the AIDE agent, we inject an `oumi_helper.py` into its workspace at runtime. The agent calls `run_trial()`, `test_reward()`, or `test_eval()` — these handle config loading from YAML, environment setup, and metric extraction. This works with any model, trainer, or config.

### Dependency handling

`aideml` pins exact versions of pandas/numpy/scipy that conflict with Oumi's requirements, but the code runs fine with newer versions. For `uv` users, `override-dependencies` in `pyproject.toml` resolves this automatically. For `pip` users, the workaround is documented in the extra definition.

### What's included

**Source** — 17 new files following Oumi conventions (copyright headers, dataclass configs, try/except optional imports, builder pattern, device cleanup, distributed support, telemetry).

**Tests** — 59 tests across 3 files:
- `test_aide_params.py` — param validation, YAML roundtrip, CLI overrides
- `test_aide_optimizer.py` — task descriptions for all 4 surfaces, config conversion, optimizer lifecycle
- `test_cli_aide.py` — CLI with mocked AIDE runs

**Notebooks** — 3 tutorials:
- **AIDE Agentic Optimization** — CONFIG_SEARCH, the main tutorial
- **AIDE Reward Function Design** — tests existing reward as baseline, lets AIDE redesign it
- **AIDE Custom Evaluation** — dataset exploration, eval pipeline breakdown, AIDE generation

**Config and examples:**
- `configs/recipes/smollm/aide/135m/aide.yaml` with `smollm-135m` alias
- `scripts/examples/aide/run_aide_optimization.py`

### How to test

```bash
# Install
uv pip install -e ".[aide]"

# Unit tests (no GPU or API key needed)
pytest tests/unit/core/configs/params/test_aide_params.py tests/unit/core/agentic/test_aide_optimizer.py tests/unit/cli/test_cli_aide.py -v

# CLI
oumi aide --help

# End-to-end (requires LLM access + GPU)
oumi aide -c smollm-135m --aide.steps=5

```

## Related issues
New feature; no existing issue.

## Before submitting
- [ ] This PR only changes documentation.
- [x] Did you read the [contributor guideline](https://github.com/oumi-ai/oumi/blob/main/CONTRIBUTING.md) Pull Request guidelines?
- [ ] Did you link the issue(s) related to this PR in the section above?
- [x] Did you add / update tests where needed?

### CODE REVIEW

**Comment by oelachqar** [comment]
Hi @efsiatras, 

Thank you so much for the contribution! Were you able to use the integration? It would be interesting to have some results demonstrating the effectiveness for LLM fine-tuning

**Comment by efsiatras** [comment]
Hi @oelachqar, thanks for the look.

Yes, I tested the integration while building it. That's what the notebooks and the recipe came out of. On effectiveness more broadly, the papers I linked in the PR description (MLE-bench, LLM Speedrunning, RE-Bench, AI Scientist-v2) cover how the algorithm works and its results across a wide range of ML tasks.

I haven't put together is a baseline-vs-AIDE results table on an Oumi fine-tuning run specifically, since I wasn't sure what workflow or dataset you'd want it on. If you let me know the workflow(s) and rough scale you'd accept, I'll run it and post the results here.

**Comment by efsiatras** [comment]
Following up on this @oelachqar.
